### DISCUSSION TITLE
How can I use Arm64DDC.swift

### DISCUSSION THREAD

**Discussion by RayPS** [MonitorControl/MonitorControl#1104] [Q&A] [answered]
Hello, I came across from https://github.com/waydabber/m1ddc/
According to the readme, since `m1ddc` can not be use as a library, I'm very interesting in [Arm64DDC.swift](https://github.com/MonitorControl/MonitorControl/blob/main/MonitorControl/Support/Arm64DDC.swift).
I'm trying to simply read the display brightness in DDC, I have look through `Arm64DDC.swift` but I still have no idea how to use this file.
I assume I can just use `read(service,command)` right?

```swift
let service = IOAVService? // How do I initialize an IOAVService?
print(Arm64DDC.read(service: service, command: 0x10)) // Read Brightness
```
I also had the [Bridging-Header.h](https://github.com/MonitorControl/MonitorControl/blob/main/MonitorControl/Support/Bridging-Header.h) included.

**Comment by waydabber** [comment, marked as answer]
Hi there,

initializing the service is somewhat complicated as you need to figure out which service belongs to which display. You need to do it in a way that even if multiple identical displays are present, you still have a service for each display. You can check out how this works in MonitorControl by starting from here:

https://github.com/MonitorControl/MonitorControl/blob/42bf5fdcf45269b4fb9e9ee08b78dd7d0ddeba95/MonitorControl/Support/DisplayManager.swift#L304

**Comment by RayPS** [comment]
@waydabber  Thanks for the help! After I reading through `DisplayManager.swift`, I believe `Arm64DDC.getServiceMatches` is what I should take care of, but one thing I found weird, when I try to use this function it always get nothing, but when I clone the MonitorControl project and remove all the unrelated files and code, it does returns a `IOAVService`.
I tested it by creating a new empty project in Xcode, added `Arm64DDC.swift`,`Bridging-Header.h`, `CoreDisplay.framework`, then I'm getting NULL, what am I missing?
![Screenshot 2022-07-08 at 16 07 47@2x](https://user-images.githubusercontent.com/1941540/177948635-54000e80-30e6-449b-b21f-1f3bbb7b25ba.png)
![Screenshot 2022-07-08 at 16 06 40@2x](https://user-images.githubusercontent.com/1941540/177948651-95890fc5-5cbe-491b-a58a-482096aa5d36.png)

**Reply by RayPS** [reply]
Oops never mind, turns out it was the App Sandbox

**Reply by waydabber** [reply]
All right. :) Good luck with the project!
